File: CS120/assignments/02/pi.c
/*****************************************************************************
filename    pi.c
author      Marcus Meng
DP email    marcus.meng@digipen.edu
course      CS120
section     A
assignment  2
due date    October 18, 2009

Brief Description:
  This file contains two functions and any associated code necessary to
  approximate the value of pi using the area of a circle and the Leibniz
  method.
*****************************************************************************/

#include <math.h> /* Need square root function. */

/* The radius of the circle we are using for the approximation is always 1.0.
 */
#define RADIUS 1.0

/* Calculates the height of a right-angle triangle given two measurements.
 *
 * base^2 + height^2 = hypotenuse^2, so height can be solved for.
 */
double height(double base, double hypotenuse)
{
    return(sqrt(hypotenuse*hypotenuse - base*base));
}

/* Approximates pi using the area of a circle.
 *
 * Compute the area with a sort of Riemann Sum, then solve for pi using the
 * equation A = pi * r^2.
 *
 * We are using the midpoint of the top of the rectangle for the
 * approximation.
 *
 * Note that we're only calculating a quarter circle's area. Final area
 * needs to be multiplied by 4.
 */
double circle_pi(int rectangles)
{
    int i;
	double sum = 0.0, width = RADIUS/rectangles;

	for (i = 0; i < rectangles; ++i)
	{
        sum += width * height((0.5+i)*width, RADIUS);
	}

    return(4.0 * sum);
}

/* Approximates pi using Leibniz's method.
 *
 * General series is pi/4 = 1 - 1/3 + 1/5 - 1/7...
 */
double leibniz_pi(int iterations)
{
	int denominator = 1;
	double sum = 0.0;

	while (denominator < iterations*2)
	{
	    if ((denominator/2) % 2 == 0)
		{
	        sum += 1.0 / denominator;
	    }
		else
		{
		    sum += -1.0 / denominator;
		}
		
		denominator += 2;
	}

    return(sum * 4);
}

File: cs120/assignments/02/erlang/pi.erl
-module(pi).
-export([circle_pi/1, leibniz_pi/1, main/0]).

%% We're using a unit circle for all the calculations here.
-define(radius, 1.0).

%% Just to help get the height of a right-angle triangle.
height(Base, Hypotenuse) when Base*Base =< Hypotenuse * Hypotenuse ->
	math:sqrt(Hypotenuse*Hypotenuse - Base*Base);
height(_, _) ->
	0.

%% Calculates pi by approximating it via the area of a circle formula.
circle_pi(Rectangles) ->
	circle_pi(?radius/Rectangles, 0.0, 0.0).

circle_pi(Width, Length, Sum) when Length < 1 ->
	circle_pi(Width, Length+Width, Sum + Width*height(Width/2 + Length, ?radius));
circle_pi(_, _, Sum) ->
	4*Sum.

%% Approximates pi using Leibniz's method.
leibniz_pi(Iterations) ->
	leibniz_pi(Iterations, 1, 0).

leibniz_pi(Iterations, Denominator, Sum) when Denominator < (Iterations*2) ->
	case ((Denominator div 2) rem 2) of
		0 -> leibniz_pi(Iterations, Denominator+2, Sum + 1/Denominator);
		1 -> leibniz_pi(Iterations, Denominator+2, Sum - 1/Denominator)
	end;
leibniz_pi(_, _, Sum) ->
	4*Sum.

%% Prints out all the values of pi calculated nicely.
print_values() ->
	print_values(1).

print_values(Current) when Current =< 1000000 ->
	io:format("~10B~20.12f~16.12f~n", [Current, circle_pi(Current), leibniz_pi(Current)]),
	print_values(Current*10);
print_values(_) ->
	ok.

main() ->
	io:format("Approximations for pi~n"),
	io:format("Iterations      Circle Method   Leibniz Method~n"),
	io:format("----------------------------------------------~n"),
	print_values().

